providers/mlx5/dr_ste.c

```diff
@@ -0,0 +1,2409 @@
+			  in_out##_first_mpls_label);\
+	DR_STE_SET_MASK_V(lookup_type, mask, mpls0_s_bos, mask, \
+			  in_out##_first_mpls_s_bos); \
+	DR_STE_SET_MASK_V(lookup_type, mask, mpls0_exp, mask, \
+			  in_out##_first_mpls_exp); \
+	DR_STE_SET_MASK_V(lookup_type, mask, mpls0_ttl, mask, \
+			  in_out##_first_mpls_ttl); \
+} while (0)
+
+#define DR_STE_SET_MPLS_TAG(lookup_type, mask, in_out, tag) do { \
+	DR_STE_SET_TAG(lookup_type, tag, mpls0_label, mask, \
+		       in_out##_first_mpls_label);\
+	DR_STE_SET_TAG(lookup_type, tag, mpls0_s_bos, mask, \
+		       in_out##_first_mpls_s_bos); \
+	DR_STE_SET_TAG(lookup_type, tag, mpls0_exp, mask, \
+		       in_out##_first_mpls_exp); \
+	DR_STE_SET_TAG(lookup_type, tag, mpls0_ttl, mask, \
+		       in_out##_first_mpls_ttl); \
+} while (0)
+
+#define DR_STE_IS_OUTER_MPLS_OVER_GRE_SET(_misc) (\
+	!!_misc->outer_first_mpls_over_gre_label || \
+	!!_misc->outer_first_mpls_over_gre_exp || \
+	!!_misc->outer_first_mpls_over_gre_s_bos || \
+	!!_misc->outer_first_mpls_over_gre_ttl)
+#define DR_STE_IS_OUTER_MPLS_OVER_UDP_SET(_misc) (\
+	!!_misc->outer_first_mpls_over_udp_label || \
+	!!_misc->outer_first_mpls_over_udp_exp || \
+	!!_misc->outer_first_mpls_over_udp_s_bos || \
+	!!_misc->outer_first_mpls_over_udp_ttl)
+
+#define DR_STE_CALC_LU_TYPE(lookup_type, rx, inner) \
+	((inner) ? DR_STE_LU_TYPE_##lookup_type##_I : \
+		   (rx) ? DR_STE_LU_TYPE_##lookup_type##_D : \
+			  DR_STE_LU_TYPE_##lookup_type##_O)
+
+enum dr_ste_tunl_action {
+	DR_STE_TUNL_ACTION_NONE		= 0,
+	DR_STE_TUNL_ACTION_ENABLE	= 1,
+	DR_STE_TUNL_ACTION_DECAP	= 2,
+	DR_STE_TUNL_ACTION_L3_DECAP	= 3,
+};
+
+enum dr_ste_action_type {
+	DR_STE_ACTION_TYPE_ENCAP_L3	= 3,
+	DR_STE_ACTION_TYPE_ENCAP	= 4,
+};
+
+static uint32_t dr_ste_crc_tab32[8][256];
+
+struct dr_hw_ste_format {
+	uint8_t ctrl[DR_STE_SIZE_CTRL];
+	uint8_t tag[DR_STE_SIZE_TAG];
+	uint8_t mask[DR_STE_SIZE_MASK];
+};
+
+static inline void dr_ste_calc_lookup_entry(uint32_t (*tbl)[256], uint8_t i,
+					    uint8_t j)
+{
+	tbl[i][j] = (tbl[i-1][j] >> 8) ^ tbl[0][tbl[i-1][j] & 0xff];
+}
+
+void dr_ste_crc32_init_table(void)
+{
+	uint32_t crc, i, j;
+
+	for (i = 0; i < 256; i++) {
+		crc = i;
+		for (j = 0; j < 8; j++) {
+			if (crc & 0x00000001L)
+				crc = (crc >> 1) ^ DR_STE_CRC_POLY;
+			else
+				crc = crc >> 1;
+		}
+		dr_ste_crc_tab32[0][i] = crc;
+	}
+
+	/* Init CRC lookup tables according to crc_slice_8 algorithm */
+	for (i = 0; i < 256; i++) {
+		dr_ste_calc_lookup_entry(dr_ste_crc_tab32, 1, i);
+		dr_ste_calc_lookup_entry(dr_ste_crc_tab32, 2, i);
+		dr_ste_calc_lookup_entry(dr_ste_crc_tab32, 3, i);
+		dr_ste_calc_lookup_entry(dr_ste_crc_tab32, 4, i);
+		dr_ste_calc_lookup_entry(dr_ste_crc_tab32, 5, i);
+		dr_ste_calc_lookup_entry(dr_ste_crc_tab32, 6, i);
+		dr_ste_calc_lookup_entry(dr_ste_crc_tab32, 7, i);
+	}
+}
+
+/* Compute CRC32 (Slicing-by-8 algorithm) */
+static uint32_t dr_crc32_slice8_calc(const void *input_data, size_t length)
```

Comment:
Is this a generic crc32? Would be happier to just use standard crc stuff from ccan:

https://github.com/rustyrussell/ccan/tree/master/ccan/crc